edb/common/ast/base.py

```diff
@@ -219,24 +232,17 @@ def __init__(self, **kwargs):
                 f'cannot instantiate abstract AST node '
                 f'{self.__class__.__name__!r}')
 
-        should_check_types = __debug__ and _check_type is _check_type_real
-        for field_name, field in self.__class__._fields.items():
-            if field_name in kwargs:
-                value = kwargs[field_name]
-            elif field.factory is not None:
-                value = field.factory()
-            else:
-                value = field.default
+        # Make kwargs directly into our __dict__
+        for field_name, factory in self._field_factories:
+            if field_name not in kwargs:
+                kwargs[field_name] = factory()
 
-            if should_check_types:
-                self.check_field_type(field, value)
+        should_check_types = __debug__ and _check_type is _check_type_real
+        if should_check_types:
+            for k, v in kwargs.items():
+                self.check_field_type(self._fields[k], v)
```

Comment:
OT: now that this all is properly typed, should we get rid of runtime type checking code altogether?  It's behind a debug flag anyway.